stan/math/prim/mat/fun/ocl_gpu.hpp

```diff
@@ -0,0 +1,296 @@
+ * platforms/devices and initialize the OpenCL
+ * context/command queue.
+ *
+ */
+class ocl {
+ private:
+  std::string description_;
+  cl::Context oclContext_;
+  cl::CommandQueue oclQueue_;
+  cl::Platform oclPlatform_;
+  cl::Device oclDevice_;
+
+  size_t max_workgroup_size;
+
+  inline void init() {
+    try {
+      std::vector<cl::Platform> allPlatforms;
+      cl::Platform::get(&allPlatforms);
+      if (allPlatforms.size() == 0) {
+        domain_error("OpenCL Initialization", "[Platform]", "",
+                     "No OpenCL platforms found");
+      }
+      oclPlatform_ = allPlatforms[0];
+      std::vector<cl::Device> allDevices;
+      oclPlatform_.getDevices(DEVICE_FILTER, &allDevices);
+      // TODO(Steve): This should throw which platform
+      if (allDevices.size() == 0) {
+        domain_error("OpenCL Initialization", "[Device]", "",
+                     "No OpenCL devices found on the selected platform.");
+      }
+      oclDevice_ = allDevices[0];
+      description_ = "Device " + oclDevice_.getInfo<CL_DEVICE_NAME>()
+                     + " on the platform "
+                     + oclPlatform_.getInfo<CL_PLATFORM_NAME>();
+      allDevices[0].getInfo<size_t>(CL_DEVICE_MAX_WORK_GROUP_SIZE,
+                                    &max_workgroup_size);
+      oclContext_ = cl::Context(allDevices);
+      oclQueue_ = cl::CommandQueue(oclContext_, oclDevice_,
+                                   CL_QUEUE_PROFILING_ENABLE, NULL);
+      init_kernel_groups();
+      // Compile the dummy kernel used for timing purposes
+      cl::Program::Sources source(
+          1, std::make_pair(dummy_kernel.c_str(), dummy_kernel.size()));
+      cl::Program program_ = cl::Program(oclContext_, source);
+
+      try {
+        program_.build(allDevices);
+        kernels["dummy"] = cl::Kernel(program_, "dummy", NULL);
+        compiled_kernels["timing"] = true;
+      } catch (const cl::Error &e) {
+        domain_error(
+            "OpenCL Initialization", e.what(), e.err(),
+            "\nRetrieving build log\n",
+            program_.getBuildInfo<CL_PROGRAM_BUILD_LOG>(allDevices[0]).c_str());
+      }
+    } catch (const cl::Error &e) {
+      check_ocl_error("build", e);
+    }
+  }
+
+ public:
+  inline std::string description() const {
+    if (initialized_ == 1) {
+      return description_;
+    }
+    return "No device selected yet.";
+  }
+
+  inline cl::Context &context() {
+    if (initialized_ == 0) {
+      init();
+      initialized_ = 1;
+    }
+    return oclContext_;
+  }
+
+  inline cl::CommandQueue &queue() {
+    if (initialized_ == 0) {
+      init();
+      initialized_ = 1;
+    }
+    return oclQueue_;
+  }
+  inline int maxWorkgroupSize() { return max_workgroup_size; }
+};
+
+static ocl ocl_context_queue;
+
+/**
+ * Returns the description of the OpenCL
+ * platform and device that is used.
+ *
+ */
+inline std::string get_description() { return ocl_context_queue.description(); }
```

Comment:
use member function instead

Reply:
use member function as in "We should just use `ocl_context_queue.description()` directly or "Make this a member function of the class"?